packages/backend-common/src/reading/AwsS3UrlReader.ts

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import aws, { Credentials, S3 } from 'aws-sdk';
+import { CredentialsOptions } from 'aws-sdk/lib/credentials';
+import {
+  ReaderFactory,
+  ReadTreeResponse,
+  ReadUrlOptions,
+  ReadUrlResponse,
+  SearchResponse,
+  UrlReader,
+} from './types';
+import getRawBody from 'raw-body';
+import { AwsS3Integration, ScmIntegrations } from '@backstage/integration';
+
+const parseURL = (
+  url: string,
+): { path: string; bucket: string; region: string } => {
+  let { host, pathname } = new URL(url);
+
+  /**
+   * Removes the trailing '/' from the pathname to be processed
+   * as a parameter by AWS S3 SDK getObject method.
+   */
+  pathname = pathname.substr(1);
+
+  /**
+   * Checks that the given URL is a valid S3 object url.
+   * Format of a Valid S3 URL: https://bucket-name.s3.Region.amazonaws.com/keyname
+   */
+  const validHost = new RegExp(
+    /^[a-z\d][a-z\d\.-]{1,61}[a-z\d]\.s3\.[a-z\d-]+\.amazonaws.com$/,
```

Comment:
So in practice this only supports hosted amazonaws.com S3. I think there were some discussions about wanting to support other S3 compatible solutions, of which there are several. But I'd be open for getting this shipped first, and then opening up for properly handling both

(1) the host config parameter, and
(2) URLs that are not on the form where the bucket is in the host name, but rather is the first path segment instead